source/language/openpulse.rst

```diff
@@ -0,0 +1,317 @@
+
+   for p in [0 : points-1] {
+     // The key line: pick the frequency to sample
+     driveframe.frequency = start + (end-start) * p / points;
+     output[p] = 0;
+
+     for i in [0 : shots-1] {
+       // Assumes suitable calibrations for reset, x, and measure_iq
+       // and that the x gate references the same "drive" frame
+       reset q;
+       x q;
+       measure_iq q -> iq;
+
+       output[p] = (output[p] * i + iq) / (i + 1);
+     }
+   }
+
+Manipulating frames based on the state of other frames is also permitted:
+
+.. code-block:: none
+
+   // Swap phases between two frames
+   const temp = frame1.phase;
+   frame1.phase = frame2.phase;
+   frame2.phase = temp;
+
+Pulses
+------
+
+Pulses have two representations:
+
+- An array of complex samples which define the points for the pulse envelope
+- A``pulse`` type, which describes an abstract mathematical function
+  representing a pulse. This will later be materialized into a list of complex
+  samples, either by the compiler or the hardware using the parameters provided
+  to the pulse template.
+
+A value of type ``pulse`` is retrieved by calling one of the built-in pulse
+template functions. Note that each of these functions takes a type ``length``
+as a first argument, since pulses need to have a definite length. Using the
+hardware dependent ``dt`` unit is recommended, since the compiler may need to
+down-sample a higher precision pulse to physically realize it.
+
+.. code-block:: none
+
+   // amp is pulse amplitude at center
+   // center is the mean of pulse
+   // sigma is the standard deviation of pulse
+   gaussian(length:l, complex[float[32]]:amp, length:center, length:sigma)
+
+   // amp is pulse amplitude at center
+   // center is the mean of pulse
+   // sigma is the standard deviation of pulse
+   sech(length:l, complex[float[32]]:amp, length:center, length:sigma)
+
+   // amp is pulse amplitude at center
+   // center is the mean of pulse
+   // square_width is the width of the square pulse component
+   // sigma is the standard deviation of pulse
+   gaussian_square(length:l, complex[float[32]]:amp, length:center, length:square_width, length:sigma)
+
+   // amp is pulse amplitude at center
+   // center is the mean of pulse
+   // sigma is the standard deviation of pulse
+   // beta is the Y correction amplitude, see the DRAG paper
+   drag(length:l, complex[float[32]]:amp, length:center, length:sigma, float[32]:beta)
+
+   // Define a constant pulse of length l
+   constant(l:length)
+
+Play instruction
+----------------
+
+Pulses are scheduled using the ``play`` instruction. These instructions may
+only appear inside a ``defcal`` block!
+
+Play instructions have two required parameters:
+
+- a value of type ``pulse`` representing the pulse envelope
+- the frame to use for the pulse
+
+.. code-block:: none
+
+   // Play a 3 sample pulse on qubit 0's "drive" frame
+   play([1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j]) frameof("drive", %0);
```

Comment:
Seems a bit odd to me that the frame is not within the `play` parentheses. Was there any particular rationale for doing this?